Sources/PackageGraph/Pubgrub.swift

```diff
@@ -937,30 +963,47 @@ public final class PubgrubDependencyResolver {
         try run()
 
         let decisions = solution.assignments.filter { $0.isDecision }
-        var finalAssignments: [(container: PackageReference, binding: BoundVersion)] = try decisions.compactMap { assignment in
-            guard assignment.term.package != root else {
-                return nil
+        var flattenedAssignments: [PackageReference: (binding: BoundVersion, products: ProductFilter)] = [:]
+        for assignment in decisions {
+            if assignment.term.node == root {
+                continue
             }
 
-            var boundVersion: BoundVersion
+            let boundVersion: BoundVersion
             switch assignment.term.requirement {
             case .exact(let version):
                 boundVersion = .version(version)
             case .range, .any, .empty, .ranges:
                 fatalError("unexpected requirement value for assignment \(assignment.term)")
             }
 
-            let container = try provider.getContainer(for: assignment.term.package)
+            let products = assignment.term.node.productFilter()
+
+            let container = try provider.getContainer(for: assignment.term.node.package)
             let identifier = try container.packageContainer.getUpdatedIdentifier(at: boundVersion)
 
-            return (identifier, boundVersion)
+            if var existing = flattenedAssignments[identifier] {
+                assert(existing.binding == boundVersion, "Two products in one package resolved to different versions: \(existing.products)@\(existing.binding) vs \(products)@\(boundVersion)")
+                existing.products.formUnion(products)
+                flattenedAssignments[identifier] = existing
+            } else {
+                flattenedAssignments[identifier] = (binding: boundVersion, products: products)
+            }
+        }
+        var finalAssignments: [(
```

Comment:
Can we reuse the `Binding` typealias here?

Reply:
Done.